### PR TITLE
DEVPROD-4768: prioritize non-task hosts for host status job

### PR DESCRIPTION
**Pull request by Kimchelly**
DEVPROD-4768

### Description
Follow-up to #7597 - the previous change does make the job faster. Unfortunately, there's still an issue where the job only checks the first N oldest hosts. If a lot of hosts were started recently _and_ they're taking a long time to actually boot up, the job will not be able to check hosts that were started later on, which can starve out later hosts that managed to start up quickly. This doesn't cause an issue for task hosts, because they can still run tasks even if they're stuck in "starting". However, host.create and spawn hosts need to go from "starting" to "running" ASAP or else the user can't get the DNS name for their host. To work around this (and given the fact that the number of host.create/spawn hosts starting up is usually small), I changed the sorting so that:

* Spawn hosts and host.create hosts always get checked first, followed by task hosts.
* Spawn hosts and host.create hosts have a check limit that's separate from task hosts. So if the check limit is 100, Evergreen can check at most 100 spawn/host.create hosts, and 100 task hosts.

### Testing
* Updated unit tests.
* Tested in staging to verify that the host status job still works and hosts are eventually running.

### Documentation
N/A

### CODE REVIEW

File: model/host/db.go [hunk 1]
```diff
@@ -1238,22 +1238,85 @@ var (
 	awsSecretKey = bsonutil.MustHaveTag(EC2ProviderSettings{}, "Secret")
 )
 
-func StartingHostsByClient(ctx context.Context, limit int) (map[ClientOptions][]Host, error) {
+// FindStartingHostsByClient returns a map of cloud provider client options to hosts
+// with those client options that are starting up. The limit limits the number
+// of hosts that can be returned. The limit is applied separately for task hosts
+// and spawn hosts/host.create hosts.
+func FindStartingHostsByClient(ctx context.Context, limit int) ([]HostsByClient, error) {
```
Comment (hadjri) [inline model/host/db.go:1245]: Can we make 500 a default const at the top of this file?

File: model/host/db.go [hunk 2]
```diff
@@ -1238,22 +1238,85 @@ var (
 	awsSecretKey = bsonutil.MustHaveTag(EC2ProviderSettings{}, "Secret")
 )
 
-func StartingHostsByClient(ctx context.Context, limit int) (map[ClientOptions][]Host, error) {
+// FindStartingHostsByClient returns a map of cloud provider client options to hosts
```
Comment (hadjri) [inline model/host/db.go:1241]: This function now returns a list so we'll need to update this comment

File: model/host/db.go [hunk 3]
```diff
@@ -1238,22 +1238,85 @@ var (
...
+	if err != nil {
+		return nil, errors.Wrap(err, "finding starting non-task hosts")
+	}
+
+	taskHostsByClient, err := findStartingTaskHosts(ctx, limit)
+	if err != nil {
+		return nil, errors.Wrap(err, "finding starting task hosts")
+	}
+
+	hostsByClient := []HostsByClient{}
```
Comment (hadjri) [inline model/host/db.go:1260]: I think these 3 lines can be replaced with `hostsByClient := append(nonTaskHosts, taskHostsByClient...)`

**Review by hadjri — CHANGES_REQUESTED** [review]

**Comment by hadjri** [comment]
> Spawn hosts and host.create hosts always get checked first, followed by task hosts.
Is this still worth highlighting if there are now separate limits in place?

**Review by Kimchelly — COMMENTED** [review]
Our specific internal algorithms aren't really something general users need to know since it doesn't change the behavior of host.list/host.create, only the performance of them. I think the general understanding is that host.create/host.list should be as fast as we can reasonably make them, and I don't really think users otherwise care what machinery Evergreen has to achieve that.

**Review by hadjri — APPROVED** [review]

**Pull request merged**